docs/madr/decisions/056-policy-matching-with-real-resources.md

```diff
@@ -0,0 +1,548 @@
+we're going to iterate over `Rules` and check if our outbound belongs to the `Rules[].Subset`.
+
+The approach has a number of disadvantages:
+
+* even though it works well for "from" policies, using it for "to" policies doesn't feel natural
+  * there is no concept of "subsets" in outbounds, we have to artificially build `{kuma.io/service: $svc}` subset
+  * subset-based matching is build to handle subset overlapping scenarios, but it's never happening in outbounds
+* when referencing the real MeshHTTPRoute we had to use subset with artificial `__rule-matches-hash__` key
+* adding support for `MeshService` forces us to introduce another artificial key like `__resource-name__`
+* Inspect API with subsets is tricky when referencing real resources (like MeshHTTPRoute or MeshService)
+
+We need a new format that'd make things easier when it comes to referencing real resources in `to[]`.
+
+## Considered Options and Decision Outcome
+
+No changes for "from" policies, both policy matching algorithm and Inspect API stays the same.
+
+For "to" policies the algorithm is going to return the additional structure:
+
+```go
+type ToRules struct {
+    Rules         Rules 
+    ResourceRules map[UniqueResourceKey]ResourceRule // new field
+}
+
+// UniqueResourceKey is a way to uniquely identify a resource, should include ResourceType, Name, Mesh and SectionName (for MeshServices)
+type UniqueResourceKey struct{}
+
+type ResourceRule struct {
+    Resource            core_model.ResourceMeta
+	ResourceSectionName string
+    Conf                interface{}
+    Origin              []Origin
+}
+
+type Origin struct {
+    Resource core_model.ResourceMeta
+    // ToIdx is an index in the 'to[]' array, so we could unambiguously detect what to-item contributed to the final conf.
+    // Especially useful when to-item uses `targetRef.Labels`, because there is no obvious matching between the specific resource
+    // in `ResourceRule.Resource` and to-item.
+    ToIdx int 
+}
+```
+
+For a simple targetRef with `name/namespace`:
+
+```yaml
+to:
+  - targetRef:
+      kind: MeshService
+      name: backend
+      namespace: finance
+    conf: conf1
+```
+
+we're going to produce a single `ResourceRule` entity:
+
+```yaml
+resourceRules:
+  meshservice:mesh/mesh-1:name/backend:ns/finance:
+    resource:
+      name: backend.finance
+      mesh: mesh-1
+    conf: conf-1
+    origin: [...]
+```
+
+When targetRef has kind `MeshService` and uses `labels` we're going to produce multiple `ResourceRule` entities
+for each real `MeshService` that satisfies `labels`, i.e for the given `targetRef`:
+
+```yaml
+to:
+  - targetRef:
+      kind: MeshService
+      labels:
+        k8s.kuma.io/namespace: finance
+    conf: conf1
+```
+
+we are going to fetch all `MeshService` resources, filter them by `k8s.kuma.io/namespace: finance` and produce N rules:
+
+```yaml
+resourceRules:
+  meshservice:mesh/mesh-1:name/finance-backend:ns/finance:
```

Comment:
While the internal structure can be a dictionnary it's should be transformed into an array in the API